pkg/docker/config/config.go

```diff
@@ -606,6 +636,45 @@ func findAuthentication(registry, path string, legacyFormat bool) (types.DockerA
 	return types.DockerAuthConfig{}, nil
 }
 
+func findAuthenticationForRef(auths dockerConfigFile, ref reference.Named) (types.DockerAuthConfig, error) {
+	for _, registry := range pathsForRef(ref) {
+		normalizedAuth := normalizeRegistry(registry)
+		if val, exists := auths.AuthConfigs[normalizedAuth]; exists {
+			return decodeDockerAuth(val)
+		}
+	}
+
+	return types.DockerAuthConfig{}, nil
+}
+
+// pathsForRef returns the valid paths for a provided reference. For example,
+// when given a reference "quay.io/repo/ns/image:tag", then it would return
+// - quay.io/repo/ns/image
+// - quay.io/repo/ns
+// - quay.io/repo
+// - quay.io
+func pathsForRef(ref reference.Named) []string {
+	// The image without tag or digest.
+	paths := []string{reference.TrimNamed(ref).String()}
+
+	// All other possible paths
+	split := strings.Split(stripImageName(ref), "/")
+	for i := len(split) - 1; i >= 0; i-- {
+		paths = append(paths, strings.Join(split[:i+1], "/"))
+	}
+
+	return paths
+}
+
+// stripImageName removes the image name from the provided ref.
+func stripImageName(ref reference.Named) string {
+	splitRefPath := strings.Split(reference.Path(ref), "/")
+	return strings.Join(append(
+		[]string{reference.Domain(ref)},
+		splitRefPath[:len(splitRefPath)-1]...,
+	), "/")
+}
```

Comment:
This can be expressed simpler (and, much less important, a bit more efficient), at the very least not using a different approach for the initial `TrimNamed` vs. `Domain+Path`, and the `splitRefPath` array seems also unnecessary.

 Compare `docker/policyconfiguration.DockerReferenceNamespaces`.

Reply:
Thank you for the hint, changed as suggested and it's now much simpler.